index.js

```diff
@@ -76,72 +67,77 @@ module.exports = class ServeDrive extends ReadyResource {
     const { pathname, searchParams } = new URL(req.url, 'http://localhost')
 
     const id = searchParams.get('drive') // String or null
-    const drive = this.drives.get(id)
-
-    if (!drive) {
-      res.writeHead(404).end('DRIVE_NOT_FOUND')
-      return
-    }
-
-    const version = searchParams.get('checkout')
-    const snapshot = version ? drive.checkout(version) : drive
-    const filename = decodeURI(pathname)
+    const drive = await this.getDrive(id)
 
-    if (this._onfilter && !this._onfilter(id, filename)) {
-      res.writeHead(404).end('ENOENT')
-      return
-    }
-
-    let entry
     try {
-      entry = await snapshot.entry(filename)
-    } catch (e) {
-      const msg = e.code || e.message
+      if (!drive) {
+        res.writeHead(404).end('DRIVE_NOT_FOUND')
+        return
+      }
 
-      if (e.code === 'SNAPSHOT_NOT_AVAILABLE') res.writeHead(404)
-      else res.writeHead(500)
+      const version = searchParams.get('checkout')
+      const snapshot = version ? drive.checkout(version) : drive
+      if (version) res.on('close', () => snapshot.close())
 
-      res.end(msg)
-      return
-    }
+      const filename = decodeURI(pathname)
 
-    if (!entry || !entry.value.blob) {
-      res.writeHead(404).end('ENOENT')
-      return
-    }
+      if (this._onfilter && !this._onfilter(id, filename)) {
```

Comment:
why not just pass the filename to getDrive also and allow that to return a falsy value if it doesn't wanna continue?

Reply:
Ah yes, that would also fulfil the filter logic. @LuKks , this ok for you? (can be follow-up PR, keet-core doesn't use filters so won't impact that code)